### ISSUE TITLE
FEATURE REQUEST : Kindly provide the gradient for tf.assign

### ISSUE THREAD

**Issue by Raj-08** [tensorflow/tensorflow#17735]
Hi, i am trying to connect a CNN to another custom layer i have created. However the custom layer that i have created requires the variables in my custom layer to be assigned from the outputs of CNN.

One may think why i need variables if its assigned from CNN at every iteration. 
Reason for variable creation is to calculate explicit gradients(via Register gradient and pyFunc) that involves sparse linear equations ,so that i can apply these gradients on my variables so that the error is back propagated all the way into CNN.

However when i call tf.gradients ,all the values are None for CNN. I found a link over stackoverflow stating that gradients are not defined for tf.assign. 

https://stackoverflow.com/questions/46984316/how-tf-assign-compute-gradients-in-tensorflow

Could you please provide gradient for tf.assign so that i can go for an end to end training.

Thanks.

### Source code / logs
     with tf.variable_scope("xyz", reuse=tf.AUTO_REUSE) as scope:        
     F = tf.get_variable("FXT",  initializer=''')
     B = tf.get_variable("BXT",  initializer='')   
     lambda_tf = tf.get_variable("lamda_tf",initializer = [100.0],dtype=tf.float32)

    
    assign_op1 = tf.assign(F,Fx)
    assign_op2 = tf.assign(B,Bx)     
    
    op=custom layer(F,B)
    grads=tf.gradients(op,tf.all_trainables())
Here Fx and Bx are values from CNN

In my backprop i have custom gradients that i have explicitly calculated and over-riden with tf.RegisterGradient and gradient overide.
When i apply calculate grads i get
[None,None,None............None{None values returned for variables in CNN},Defined Gradients{Defined Gradients returned for variables in Custom Layer}]

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by Raj-08** [comment]
Have I written custom code : Partially yes
OS Platform and Distribution : Ubuntu 16.04
TensorFlow installed from : pip
TensorFlow version : 1.4
Bazel version : N/A
CUDA/cuDNN version : 8.0/5.0
GPU model and memory : TeslaK80 , 12 GB
Exact command to reproduce : N/A

**Comment by angersson** [comment]
@vrv Can you take a look at this?

**Comment by vrv** [comment]
Assigning to Alex, who has more state on the roadmap for everything related to Variables.  Please feel free to re-assign!

**Comment by yaroslavvb** [comment]
@Raj-08 tf.assign returns "ref", ie link to the variable. So if you implemented backprop for it, the result would be zero since state of the variable is functionally independent of other parts of the graph.

Do you want the gradients to just propagate the right hand side of the assignment? If so, you could differentiate the right-hand side explicitly, and use control dependencies to force assignment to happen on forward pass

```
a = tf.Variable(1)
b = tf.assign(a, stuff)
with tf.control_dependencies([b]):
  c = tf.identity(stuff)

```

**Comment by alextp** [comment]
There is no plan to add a gradient to tf.assign because it's not possible in general to connect the uses of the assigned variable with the graph which assigned it.

Instead, your solution of setting the weights of a network by the output of another is better implemented by not using tf.Variable at all for the computed weights, and just using tensors. We currently make it not too easy to do this using the tf.layers library but you can do this if you subclass the layers in tf.layers and override the build() method.

**Comment by Raj-08** [comment]
@yaroslavvb It worked thanks!!!!

There were no errors in graph construction

But during the graph execution in backprop i get this error and its back tracked to 

`tf.gradients(c, ...)`
then it points to the
`line where i defined the "stuff"[as per your comment]`

ERROR:
`Input to reshape is a tensor with 16777216 values, but the requested shape has
4096`

Do i need to make some other changes as well?

**Comment by Raj-08** [comment]
@alextp How can apply my gradients to tensors?

**Comment by Raj-08** [comment]
@alextp  @yaroslavvb I have checked all my ops in backprop and i am making sure i am passing the gradient values of required shape. Is this an internal problem when creating custom op and gradient?

**Comment by alextp** [comment]
@Raj-08 what do you mean "apply my gradients to tensors"? Either you want to store mutable variables as your layer weights (which are initialized from the predictions of another network), in which case you apply your gradients to the variables but you can't backprop through them, or you want your layer weights to be computed tensors, in which case you should apply the gradients not to them but to the network which is producing them, right?

Or is there another option I'm missing here?

**Comment by Raj-08** [comment]
@yaroslavvb @alextp 

So my gradients are flowing back right until the bottom of the CNN(FCN) and then i am seeing an unconnected node on my tensorflow graph visualization. Even though the loss has been decreasing i am not able to find anything meaningful on the final output.


![step0001](https://user-images.githubusercontent.com/15856029/37711449-c3f0c416-2d36-11e8-8f43-5b9be2cd3eef.png)

Are the gradients still flowing through that node?

**Comment by Raj-08** [comment]
@alextp No you aren't missing anything . 

Let me tell you why i went for creating variables in the first place.

At the start i computed the CNNs outputs as tensors and didnt involve any variabe but after creating the graph and calling tf.gradients it was showing None for the custom gradients i had created. 
So i went about creating variables and then after following what @yaroslavvb mentioned i could get the gradients i have created.

After visualizing the tensorflow graph i found that all my gradients were calculated as i intended but now as in the image above we can see that there is an unconnected node in between the gradients and the FCN .






So now i have two questions :

**Comment by alextp** [comment]
I'm not sure I understand. What's the unconnected node you see?

When you just compute, on an interactive session tf.gradients on your final loss and your parameters, do you get 0 or None?

Also, if you're seeing an error you should share the whole stack trace as that's much easier to debug. The shape error you linked looks like it's a real shape error in your code.

**Comment by ahatamiz** [comment]
I am having the same issue for not being able to backprop the gradient through tf.assign. Is there any solution to this ?

**Comment by sajaddarabi** [comment]
You can try overriding the assign operator, perhaps something like the following:

`@tf.custom_gradient
def mod_assign(x):
    phi = tf.Variable(tf.zeros([1, 4], dtype='float32'), dtype='float32',  trainable=True)
    phi = tf.assign(phi,x)
    def grad(dy):
        return dy
    return phi, grad
`

**Comment by ahatamiz** [comment]
Thanks !
